extensions/big-data-cluster/src/bigDataCluster/tree/controllerTreeDataProvider.ts

```diff
@@ -0,0 +1,147 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import * as vscode from 'vscode';
+import * as azdata from 'azdata';
+import { TreeNode } from './treeNode';
+import { IControllerTreeChangeHandler } from './controllerTreeChangeHandler';
+import { AddControllerNode } from './addControllerTreeNode';
+import { ControllerRootNode, ControllerNode } from './controllerTreeNode';
+import { IEndPoint } from '../controller/clusterController';
+
+const ConfigNamespace = 'clusterControllers';
+const CredentialNamespace = 'clusterControllerCredentials';
+
+export class ControllerTreeDataProvider implements vscode.TreeDataProvider<TreeNode>, IControllerTreeChangeHandler {
+
+	private _onDidChangeTreeData: vscode.EventEmitter<TreeNode> = new vscode.EventEmitter<TreeNode>();
+	public readonly onDidChangeTreeData: vscode.Event<TreeNode> = this._onDidChangeTreeData.event;
+	private root: ControllerRootNode;
+	private credentialProvider: azdata.CredentialProvider;
+
+	constructor() {
+		this.root = new ControllerRootNode({ treeChangeHandler: this });
+		this.loadSavedControllers();
+	}
+
+	public async getChildren(element?: TreeNode): Promise<TreeNode[]> {
+		if (element) {
+			return element.getChildren();
+		}
+
+		if (this.root.hasChildren) {
+			return this.root.getChildren();
+		} else {
+			return [new AddControllerNode()];
+		}
+	}
+
+	public getTreeItem(element: TreeNode): vscode.TreeItem | Thenable<vscode.TreeItem> {
+		return element.getTreeItem();
+	}
+
+	public addController(
+		url: string,
+		username: string,
+		password: string,
+		rememberPassword: boolean,
+		masterInstance?: IEndPoint
+	): void {
+		this.root.addControllerNode(url, username, password, rememberPassword, masterInstance);
+		this.notifyNodeChanged();
+	}
+
+	public deleteController(url: string, username: string): ControllerNode {
+		let deleted = this.root.deleteControllerNode(url, username);
+		if (deleted) {
+			this.notifyNodeChanged();
+		}
+		return deleted;
+	}
+
+	public notifyNodeChanged(node?: TreeNode): void {
+		this._onDidChangeTreeData.fire(node);
+	}
+
+	public async loadSavedControllers(): Promise<void> {
+		let config = vscode.workspace.getConfiguration(ConfigNamespace);
```

Comment:
Is there a reason we can't save all this in a memento (as Azure does) instead of in the configuration? This is already very full and we should try to avoid adding more stuff there.

Reply:
Because you and @alanrenmsft gave me CMS as a guidance for building controller extension, and CMS stores them in config. Please provide this kind of input in design/planning step, otherwise I just follow the guidance you provided. If it really needs to be stored in memento, I will handle it in separate PR.

Reply:
Please create a tracking issue and let's work to address in a future PR.